### PR TITLE
Implemented Cypress as integration test framework

### PR DESCRIPTION
**Pull request by logan-hendricks**
Implemented Cypress integration testing framework into vue client.

How to run:

1. (recommended for first time) delete your current node image : `docker rmi node:12-slim`
2. rebuild the uno dockers `docker-compose up` in the root of the project
3. Once that has completely finished (it will take a bit as it rebuilds the DockerFile), run the integration test file `docker-compose -f integration-tests.yml up`

### CODE REVIEW

File: client/Dockerfile [hunk 1]
```diff
@@ -2,6 +2,19 @@ FROM node:12-slim
...
+    libnss3 \
+    libxss1 \
+    libasound2 \
+    libxtst6 \
+    xauth \
+    xvfb \
+    chromium
+
 RUN npm install -g @vue/cli && \
     npm install -g @vue/cli-service-global && \
```
Comment (logan-hendricks) [inline client/Dockerfile:19]: @Aaron-Fine To answer your question from my previous PR.... we're basically doing npm install cause that's how Jacob put it :) I didn't feel like updating it.

File: client/cypress/integration/login.spec.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    cy.visit('http://localhost:3000')
+  });
+
+  it("should join a game", () => {
+    cy.log("creating new game");
+    cy.get("#newGame").click();
+
+    cy.log("Confirm the new game is created");
+    cy.get("#status").contains("New game id is:");
+    cy.get("#gameId").should("have.value", "12234");
```
Comment (Aaron-Fine) [inline client/cypress/integration/login.spec.js:15]: It is not clear to me why the `gameID` would always be `12234`

File: client/cypress/integration/login.spec.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+/// <reference types="cypress" />
+
+context('Login', () => {
```
Comment (Aaron-Fine) [inline client/cypress/integration/login.spec.js:3]: Shouldn't this be in a `describe` and not a `context`?

File: client/src/components/Login.vue [hunk 4]
```diff
@@ -12,11 +12,11 @@
               </v-toolbar>
               <v-card-text>
                 <v-form>
-                  <v-text-field label="GAME ID" type="text" v-model="game_id"></v-text-field>
-                  <v-text-field label="USERNAME" type="text" v-model="user_name"></v-text-field>
-                  <v-btn @click.native="login" color="primary" :to="to">Join Game</v-btn>
-                  <v-btn color="primary" @click.native="newGame">Create new game</v-btn>
-                  <v-card v-if="status != ''">{{ status }}</v-card>
+                  <v-text-field id="gameId" label="GAME ID" type="text" v-model="game_id"></v-text-field>
```
Comment (Aaron-Fine) [inline client/src/components/Login.vue:15]: So far as I know, best practice is to not include `id`'s as it becomes very difficult to ensure uniqueness over time.

File: docker-compose.yml [hunk 5]
```diff
@@ -2,7 +2,7 @@ version: '3.6'
 
 services:
   client:
-    image: node:12-slim
+    image: vue-client:latest
```
Comment (Aaron-Fine) [inline docker-compose.yml:5]: Shouldn't these be `client` and `server` to reference the docker files we have?

File: integration-tests.yml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+version: '3.6'
```
Comment (Aaron-Fine) [inline integration-tests.yml:1]: Why do we need this file?

File: client/cypress/integration/login.spec.js (hunk 2 above)
Comment (logan-hendricks) [reply client/cypress/integration/login.spec.js:15]: so. this is one I was waiting for someone to point out. I'm currently working on a more dymanic process to strip out the game id. :) Locally, my machine always just serves up that same value. I'll be adjusting this for the future.

File: client/cypress/integration/login.spec.js (hunk 3 above)
Comment (logan-hendricks) [reply client/cypress/integration/login.spec.js:3]: hmmm... I know in Karma this is typically describe. This was a copy paste from their example file, and it included context as the keyword. However I can investigate into this one to confirm that. As is, It certainly runs.

File: client/src/components/Login.vue (hunk 4 above)
Comment (logan-hendricks) [reply client/src/components/Login.vue:15]: Aaron, I think this is a good call out. I fully recognize this is counter to what most Computer Science courses teach. For my two cents on the matter. In my experience, I fully will admit that for development purposes, I would not recommend using `ids` for styling or identifying sections of code.  

However, from a testing methodology, not including `Id's` tends to require automation engineers relying on things like `css selectors`, `classes`, or even worse `xpath`.  From that perspective, while id's do run the risk of collision, in my opnion, they are by far the more reliable process in comparison to other options.

I have done/seen approaches of creating a pre-selected arbitrary html attribute for testing suites to rely on for selecting items. I agree on that as a better long term approach, and would be happy to investigate it if you feel it's preferable for Cypress. 

In short though, I genuinely am open if you have other more preferable solution.

File: docker-compose.yml (hunk 5 above)
Comment (logan-hendricks) [reply docker-compose.yml:5]: no, that's a good question though. In docker-compose, if you define a `build` stage on a service, the `image` tag is actually what we want the finished/compiled image to be **named/tagged**.

File: integration-tests.yml (hunk 6 above)
Comment (logan-hendricks) [reply integration-tests.yml:1]: This runs the integration tests in a headless manner so it can be processed by the build/test server in the pipeline.... It's somewhat pre-mature.... it's something I figured we would need on the next steps when hooking these into the pipelines.

File: client/cypress/integration/login.spec.js (hunk 3 above)
Comment (Aaron-Fine) [reply client/cypress/integration/login.spec.js:3]: I think it really depends on what the test runner is expecting. I have mostly used mocha and things built off of it, so I was expecting a `describe`. So long as it works!

File: client/src/components/Login.vue (hunk 4 above)
Comment (Aaron-Fine) [reply client/src/components/Login.vue:15]: You can do some pretty neat things with xpath! `//button[contains(text(),"play")]` is a good example of what you can't do without it (so far as I know). Reference: https://devhints.io/xpath

That said, I totally agree that having `id`'s makes writing unit tests _so_ much easier. I guess, especially for unit tests on things like Vue components, having them makes sense and its easier to ensure uniqueness on a single component. 

This is an area where I, too, am hoping others have more experience and can offer guidance.

File: docker-compose.yml (hunk 5 above)
Comment (Aaron-Fine) [reply docker-compose.yml:5]: Thank you for the explanation!

File: client/cypress/integration/login.spec.js (hunk 3 above)
Comment (logan-hendricks) [reply client/cypress/integration/login.spec.js:3]: Yeah. so reading through their documentation. here's what I found:
```
The test interface, borrowed from Mocha, provides describe(), context(), it() and specify().

context() is identical to describe() and specify() is identical to it(), so choose whatever terminology works best for you.
```

However, it does seem odd in the text example that it does a mix and match. `context` and `it`. I'll probably change it in this case to be `describe` then to match the `it` block used in mocha.

File: client/src/components/Login.vue (hunk 4 above)
Comment (logan-hendricks) [reply client/src/components/Login.vue]: ha ha ha for sure :) xpath is a pretty powerful tool. My experience has been that I'll only use it in a worst case scenario. (like what you described above). You can do some pretty stellar things with it that are practically impossible without it.

That being said. the more I thought about this. I ended up changing the `id's` to be an arbitrary HTML attribute `test-id`. and changed all of the queries to get elements to use a css selector to pull in those specific elements. I feel like this sets a better precedent for future testing, as it no longer completely ties it to the development process. (or randomly named id's)

I ended up prefixing component name in the `test-id` in order to specify which component it belongs to. 

Obviously these are all manual choices/paradigms. I equally am open to anyone else who may have a better way this can be handled.
